Add codeberg version for PR assignment

laumann opened the pull request:
This is more or less a 1-1 porting, replacing the github API calls with equivalent Codeberg ones.

The differences are mostly:

 * Labels are assigned by their IDs, so we need to map the label names to the IDs prior to assignment

 * The PR JSON objects pulled from the Codeberg API already contain all the label information necessary (including names), so there's no need to use the issues API

 * This version attempts to report which bug numbers are invalid (if any), and only link valid ones.

 * A couple of TODOs from the github version are left in

 * PRs are not yet linked to Bugzilla, because "See Also" does not allow "codeberg.org" links. See bug [964700](https://bugs.gentoo.org/964700).

laumann committed:
Add codeberg version for PR assignment

This is more or less a 1-1 porting, replacing the github API calls
with equivalent Codeberg ones.

The differences are mostly:

 * Labels are assigned by their IDs, so we need to map the label names
 to the IDs prior to assignment

 * The PR JSON objects pulled from the Codeberg API already contain
 all the label information necessary (including names), so there's no
 need to use the issues API

 * This version attempts to report which bug numbers are invalid (if
 any), and only link valid ones.

 * A couple of TODOs from the github version are left in

 * PRs are not yet linked to Bugzilla, because "See Also" does not
 allow "codeberg.org" links. See bug 964700.

Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>

mgorny commented on codebergapi.py:
```diff
@@ -0,0 +1,80 @@
+        return f"https://codeberg.org/api/v1/repos/{self.owner}/{self.repo}"
+
+    def _get(self, url, params={}):
+        resp = requests.get(url, params=params, headers=self.headers)
```
Why not create a `requests.Session` in `__init__`?

mgorny commented on codebergapi.py:
```diff
@@ -0,0 +1,80 @@
+import requests
```
Could you add typing, do a `mypy`, `ruff check` and `ruff format`?

laumann replied:
I already did ruff check + format, the only complaints from check were about variables with short names. I can add typing, but I don't know what you mean by `mypy` :sweat_smile:

thesamesam replied:
After adding typing, run `mypy` on the source file(s) and see if it complains.

laumann replied:
mypy doesn't like how clever I was :)

```python
next_url = "..."
while next_url:
    r = self.session.get(next_url)
    yield from r.json()
    next_url = (x := r.links.get("next")) and x["url"]  # <-- this line
```
The complaint:
```
$ mypy codebergapi.py 
codebergapi.py:36: error: Incompatible types in assignment (expression has type "Any | None", variable has type "str")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

I'll fix it

mgorny commented on codebergapi.py:
```diff
@@ -0,0 +1,80 @@
+    def _get(self, url, params={}):
+        resp = requests.get(url, params=params, headers=self.headers)
+        resp.raise_for_status()
+        return json.loads(resp.content)
```
I'm pretty sure requests has a method for this.

laumann replied:
Gah, you're right - I think I knew that, but completely forgot...

laumann replied:
Done.

mgorny commented on codebergapi.py:
```diff
@@ -0,0 +1,80 @@
+            if not pulls:
+                break
+            yield from pulls
+            page += 1
```
```suggestion
        for page in itertools.count(1):
            pulls = self.pulls_page(page)
            if not pulls:
                break
            yield from pulls
```

laumann replied:
Thanks! I went with a different approach when I realised that the Codeberg API returns Link headers that requests just parses out for us. Now we won't make an extra unneeded API request only to figure out we already got the last page.

mgorny commented on codebergapi.py:
```diff
@@ -0,0 +1,80 @@
+    def create_review(self, pr_id, comment):
+        # Does not appear to be possible to simply post comments
+        # https://codeberg.org/api/swagger#/repository/repoCreatePullReview
+        url = f"{self.repos_baseurl}/pulls/{pr_id}/reviews"
```
To be honest, I really hate how you sometimes inline stuff, and sometimes introduce two variables for stuff you could inline.

laumann replied:
Fixed, I hope :D I inlined everywhere.

laumann commented on codebergapi.py:
```diff
@@ -0,0 +1,62 @@
+        }
+
+    def close(self):
+        self.session.close()
```
Was considering to implement `__enter__()` and `__exit__()` so you could do something like

```python
with CodebergAPI(owner, repo, token) as repo:
    repo.pulls()
    # ...
```
and have `__exit__()` call `self.session.close()` to properly close the requests session.

thesamesam replied:
Sure, why not? Not sure if we should then inherit from the ABC for a contextmanager or not.

laumann replied:
i've implemented `__enter__` and `__exit__` which seem to work - let me know if there's a better way to do this.

thesamesam commented on assign-pull-requests-codeberg.py:
```diff
@@ -0,0 +1,487 @@
+import re
+import lxml.etree
+
+try:
```
These seem to be some ancient compatibility hack (py2 I think)? Can't we remove the fallbacks?

laumann replied:
Removed

laumann committed:
Add type hints plus other cleanup

Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>

laumann committed:
Enable setting "See Also" URLs on bugzilla

The "See Also" field on bugs.gentoo.org now accepts codeberg.or (plus
others).

Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>

laumann committed:
Fix mypy finding, add "-> None" return type hints

Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
